### DISCUSSION TITLE
(Optional) Encryption of Fields

### DISCUSSION THREAD

**Discussion by ChuckMoe** [fief-dev/fief#75] [Ideas]
Hello together :)

the [European General Data Protection Regulation (GDPR)](https://gdpr.eu) describe multiple ways on how to enforce privacy. One is the encryption of person relatable data in the database, in case that someone gets access to it.

As @frankie567 pointed out below, encrypting the email address, that is used for login, would add complexity to the whole system. I tried to find an answer if it is still necessary but most of the post I found either cover the protection of the password or email exchange.

Nevertheless, I would welcome the option to encrypt [User Fields](https://docs.fief.dev/getting-started/user-fields/) values in the database.

== Old ==
As fas as I can see the email address is [not encrypted by default](https://github.com/fief-dev/fief/blob/45050305a841129ea30f18bf2202f072b6e9a7f7/backend/fief/models/user.py#L30).
It would be possible to encrypt the data with [sqlalchemy](https://github.com/sqlalchemy/sqlalchemy/wiki/DatabaseCrypt). I did not read through all of the fief code but as far as I can see, it only uses sqlalchemy code for working with the database (no custom sql). As far as I can tell, this should not change any other logic of fief, right?

Another option could be sqlalchemy_utils, which implements some colum types, including the [EncryptType](https://sqlalchemy-utils.readthedocs.io/en/latest/data_types.html#module-sqlalchemy_utils.types.encrypted.encrypted_type). With this you only need to set it as column type and specify parameters like the secret and it automatically encrypts/decrypts your data.
An example taken straight from the sqlalchemy_utils docs:
```
import sqlalchemy
from sqlalchemy_utils import EncryptedType
from sqlalchemy_utils.types.encrypted.encrypted_type import AesEngine

username = sqlalchemy.Column(
    EncryptedType(
        sqlalchemy.Unicode,
        secret_key,
        AesEngine,
        'pkcs5'))
```

**Comment by frankie567** [comment]
Hi @ChuckMoe 👋 

Thank you for the suggestion! Actually, we already encrypt some fields in the database, in particular:

* External databases credentials: https://github.com/fief-dev/fief/blob/45050305a841129ea30f18bf2202f072b6e9a7f7/backend/fief/models/workspace.py#L103-L108
* OAuth providers client ID and secret: https://github.com/fief-dev/fief/blob/45050305a841129ea30f18bf2202f072b6e9a7f7/backend/fief/models/oauth_provider.py#L32-L46
* OAuth accounts access and refresh tokens: https://github.com/fief-dev/fief/blob/45050305a841129ea30f18bf2202f072b6e9a7f7/backend/fief/models/oauth_account.py#L53-L72

All of them use the Fernet encryption method: https://github.com/fief-dev/fief/blob/45050305a841129ea30f18bf2202f072b6e9a7f7/backend/fief/crypto/encryption.py#L19-L26

---

As for encrypting e-mails, well, I'm not really sure of the added benefits it would have in terms of security/privacy. The e-mail address is shown in lot of places, in the Admin Dashboard and the API.

Besides, it would make login much more difficult (if not impossible) since we would not be able to query the email in the database. With the Fernet method for example, if you encrypt the same value twice, you don't get the same output.

**Reply by ChuckMoe** [reply]
Hi @frankie567 ,
thank you for the fast reply again!

You are right, I didn't think of the problem with the email query.

I must apologize for my initial answer, it didn't state the root of my concern. The European privacy regulations state that person relatable data has to be secured in some way. One option is to encrypt the data in the database in case of breaches, where the attackers get a hold of the databases. This means I do not mind when the email is shown in the admin panel or the api, only that if someone is somehow able to access the database directly, all person relatable data is encrypted.

I am going to edit my initial question to make it for clear for everyone who is coming here.

**Comment by frankie567** [comment]
Thank you for the clarifications! I think indeed encrypting e-mails is not really worth the hassle.

Encrypting user fields values, however might be interesting indeed. We could imagine two different approaches:

* Encrypt all the values by default, but I believe it would have severe performance issues when it comes to retrieve the user data.
* Have an option on User Field to explicitly toggle encryption on that field.

What do you think?

**Reply by ChuckMoe** [reply]
I also thought about having an option on what field to encrypt. Not all data is person relatable and needs encryption. :+1:

**Comment by frankie567** [comment]
By the way, thank you for making me aware of SQLAlchemy Utils `EncryptedType`: it allowed me to greatly improve the implementation! Much cleaner now: https://github.com/fief-dev/fief/commit/2ca777207a0e471f06d47bd62c75d8764297b3a6

**Comment by ChuckMoe** [comment]
If I get this right, you create one table for all different kinds of field values, right? https://github.com/fief-dev/fief/blob/2ca777207a0e471f06d47bd62c75d8764297b3a6/backend/fief/models/user_field_value.py#L20-L41

Would it not be better to create a base class field_values, then create different classes that inherit from it so that all of the types like string, int, etc. have their own table? This would keep the tables much cleaner and you can add new types relatively easily. One could try it like this
https://github.com/ChuckMoe/fief/commit/87671692d393823fecabb542f71d289cadcce4f0

**Reply by frankie567** [reply]
That's right! What you propose would work, but I purposely didn't take that path:

* Having several tables is harder to maintain, we need to account for migrations, etc.
* When retrieving the user values, we would have *N* JOIN queries to make instead of 1; which would have quite a big performance impact.

It's not the most elegant way, but I think it's a good trade-off. The DB gives us the row, and we let Python read the right column.

What I thought about this encrypted value would be to add a dedicated column `value_encrypted`; and add the necessary Python code in the model to read it and convert it to the right format.

**Reply by ChuckMoe** [reply]
Makes sense :+1: 
The first time I saw it, it simply confused me a little why you took this approach, so thanks for the explanation!
